tests/kernel/wave/attention/paged_attention_test.py

```diff
@@ -0,0 +1,253 @@
+from iree.turbine.kernel.lang.global_symbols import *
+from iree.turbine.kernel.wave.utils import (
+    get_default_run_config,
+    get_default_scheduling_params,
+    device_randn,
+    device_zeros,
+)
+from iree.turbine.kernel.wave.constraints import MMAType
+import os
+from torch.testing import assert_allclose
+from ..common.utils import (
+    require_e2e,
+    enable_scheduling_barriers,
+    dump_generated_mlir,
+)
+from ..common.shapes import get_test_shapes
+from typing import List, Optional
+
+# Reference paged attention implementation from vLLM and sglang.
+# From: https://github.com/vllm-project/vllm/blob/main/tests/kernels/test_flash_attn.py
+NUM_HEADS = [(4, 4), (8, 2), (16, 2)]
+HEAD_SIZES = [128, 256]
+BLOCK_SIZES = [16, 32]
+DTYPES = [torch.float16]
+NUM_BLOCKS = [32768, 2048]
+# First item is query length, second item is key/value length.
+# In decode, query length is always one.
+SEQ_LENS = [[(1, 523), (1, 37), (1, 2011)]]
+
+
+# From: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/attention/torch_native_backend.py
+def _run_sdpa_forward_decode(
+    query: torch.Tensor,
+    output: torch.Tensor,
+    k_cache: torch.Tensor,
+    v_cache: torch.Tensor,
+    req_to_token: torch.Tensor,
+    req_pool_indices: torch.Tensor,
+    seq_lens: torch.Tensor,
+    scaling=None,
+    enable_gqa=False,
+    causal=False,
+):
+    """Run the decode forward by using torch native sdpa op.
+
+    Args:
+        query: [num_tokens, num_heads, head_size]
+        output: [num_tokens, num_heads, head_size]
+        k_cache: [max_total_num_tokens, num_heads, head_size]
+        v_cache: [max_total_num_tokens, num_heads, head_size]
+        req_to_token: [max_num_reqs, max_context_len]
+        req_pool_indices: [num_seqs]
+        seq_lens: [num_seqs]
+        scaling: float or None
+        enable_gqa: bool
+        causal: bool
+
+    Returns:
+        output: [num_tokens, num_heads, head_size]
+    """
+
+    # [num_tokens, num_heads, head_size] -> [num_heads, num_tokens, head_size]
+    query = query.movedim(0, query.dim() - 2)
+
+    start_q, start_kv = 0, 0
+    for seq_idx in range(seq_lens.shape[0]):
+        # TODO: this loop process a sequence per iter, this is inefficient.
+        # Need optimize the performance later.
+
+        seq_len_q = 1
+        seq_len_kv = seq_lens[seq_idx]
+        end_q = start_q + seq_len_q
+        end_kv = start_kv + seq_len_kv
+
+        per_req_query = query[:, start_q:end_q, :]
+
+        # get key and value from cache. per_req_tokens contains the kv cache
+        # index for each token in the sequence.
+        req_pool_idx = req_pool_indices[seq_idx]
+        per_req_tokens = req_to_token[req_pool_idx, :seq_len_kv]
+        per_req_key = k_cache[per_req_tokens].movedim(0, query.dim() - 2)
+        per_req_value = v_cache[per_req_tokens].movedim(0, query.dim() - 2)
+
+        per_req_out = (
+            torch.nn.scaled_dot_product_attention(
```

Comment:
Shouldn't it be  `torch.nn.functional.scaled_dot_product_attention`?

Reply:
yes, will fix